### PR TITLE
Fix contains on nengo.Config.

### LINKED ISSUE #1236: Wrong type in error message
I was doing `'solver' in cfg` when I meant to do `'solver' in cfg[nengo.Connection]` and got the following error exception:
```
/home/jgosmann/.pyenv/versions/3.5.2/lib/python3.5/site-packages/nengo/config.py in __getitem__(self, key)
    342         raise ConfigError(
    343             "Type '%(name)s' is not set up for configuration. Call "
--> 344             "configures('%(name)s') first." % {'name': type(key).__name__})
    345 
    346     def __repr__(self):

ConfigError: Type 'int' is not set up for configuration. Call configures('int') first.
```
Note that it says type 'int' even though I'm testing for a string. But even with the correct type in there, it might be possible to produce a better exception.

### PR DESCRIPTION
**Pull request by jgosmann**
**Motivation and context:**
This fixes the contains function on Config objects so that one can do things like `x in cfg`. Before this could produce cryptic exceptions.

Fixes #1236.

**Interactions with other PRs:**
none

**How has this been tested?**
added a test

**How long should this take to review?**

- Quick (less than 40 lines changed or changes are straightforward)

**Types of changes:**

- Bug fix (non-breaking change which fixes an issue)

**Checklist:**

- [x] I have read the **CONTRIBUTING.rst** document.
- [n/a] I have updated the documentation accordingly.
- [ ] I have included a changelog entry.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by tbekolay** [comment]
I'm a bit confused as to the use case for this... what kind of things would I be checking is contained in a config? For the normal Nengo config, for example, am I checking that `nengo.Ensemble` is in a config, and therefore I can set ensemble options, or am I checking that an ensemble instance is in a config, and therefore that it has been configured in some way? Although we are overloading the `__getitem__` here, I don't really see `config` objects as analogous to dicts... the classes that are configurable feel markedly different from the instances of those classes.

**Comment by jgosmann** [comment]
Indeed, there is no reason to check if something in a config, but if you accidentally do (as happened to me in #1236), you would get a very confusing error.

**Comment by jgosmann** [comment]
Note that with the implementation of `__getitem__` Python provides a default implementation of `__contains__` that goes through the integer indices to check for containment (at least it seems like that is happening).

**Comment by tbekolay** [comment]
Ah, I see. I think I would prefer to raise an exception (`NotImplementedError` maybe? Suggesting the correct form in #1236?) than to return a confusing true or false. What do you think? If contentious we could talk about this at the next dev meeting!

**Comment by jgosmann** [comment]
I think formally the implementation is correct and consistent with the behaviour of `__getitem__`. There might actually be a use case because it allows to check whether a class is set up to be configured with a given `Config` instance. But it is definitely not the most self-explanatory way of checking this, so it might be better to implement this in a different way anyways if needed.

It might also be possible to raise a `TypeError` for unexpected (non-class) types and do the contains check for class types. Maybe we should do that in any case in the `__getitem__` method ([the possiblitiy of raising `TypeError` is mentioned in the Python documentation](https://docs.python.org/2/reference/datamodel.html#object.__getitem__)), but that's somewhat separate as it doesn't affect the problem of the confusing exception for `__contains__` much.

I could also see raising an exception for all `__contains__` calls as you propose. But after looking into the documentation, I think that `TypeError` might be more appropriate.
* [Type error](https://docs.python.org/2/library/exceptions.html#exceptions.TypeError): “Raised when an operation or function is applied to an object of inappropriate type.”
* [NotImplementedError](https://docs.python.org/2/library/exceptions.html#exceptions.NotImplementedError): “In user defined base classes, abstract methods should raise this exception when they require derived classes to override the method.” (`__contains__` is not supposed to be implemented by a derived class.)

Bottom line: None of the solutions appears obviously better than the others.

**Comment by tbekolay** [comment]
Sure, `TypeError` sounds good to me. Regardless, feels like it's worth chatting about this in a higher bandwidth medium, so talking about this is on the dev meeting agenda for next week.

**Comment by jgosmann** [comment]
Changed the PR to produce a TypeError instead as discussed at the dev meeting.

**Comment by jgosmann** [comment]
Technically it's a bugfix? idk, I'm probably fine either way.
